conmon/conmon.c

```diff
@@ -454,71 +497,51 @@ int main(int argc, char *argv[])
-		pexit("Failed to add stdin to epoll");
+	if (masterfd_stdout >= 0) {
+		ev.data.fd = masterfd_stdout;
+		if (epoll_ctl(epfd, EPOLL_CTL_ADD, ev.data.fd, &ev) < 0)
+			pexit("Failed to add console masterfd_stdout to epoll");
+		num_stdio_fds++;
 	}
-	*/
-	ev.data.fd = mfd;
-	if (epoll_ctl(epfd, EPOLL_CTL_ADD, mfd, &ev) < 0) {
-		pexit("Failed to add console master fd to epoll");
+	if (masterfd_stderr >= 0) {
+		ev.data.fd = masterfd_stderr;
+		if (epoll_ctl(epfd, EPOLL_CTL_ADD, ev.data.fd, &ev) < 0)
+			pexit("Failed to add console masterfd_stderr to epoll");
+		num_stdio_fds++;
 	}
 
-	/*
-	 * Log all of the container's output and pipe STDIN into it. Currently
-	 * nothing using the STDIN setup (which makes its inclusion here a bit
-	 * questionable but we need to rewrite this code soon anyway TODO).
-	 */
-	while (true) {
+	/* Log all of the container's output. */
+	while (num_stdio_fds > 0) {
 		int ready = epoll_wait(epfd, evlist, MAX_EVENTS, -1);
-		int i = 0;
-		for (i = 0; i < ready; i++) {
+		if (ready < 0)
+			continue;
+
+		for (int i = 0; i < ready; i++) {
 			if (evlist[i].events & EPOLLIN) {
-				if (evlist[i].data.fd == STDIN_FILENO) {
-					/*
-					 * TODO: We need to replace STDIN_FILENO with something
-					 *       more sophisticated so that attach actually works
-					 *       properly.
-					 */
-					num_read = read(STDIN_FILENO, buf, BUF_SIZE);
-					if (num_read <= 0)
-						goto out;
-
-					if (write(mfd, buf, num_read) != num_read) {
-						nwarn("partial/failed write (masterFd)");
-						goto out;
-					}
-				} else if (evlist[i].data.fd == mfd) {
-					num_read = read(mfd, buf, BUF_SIZE);
-					if (num_read <= 0)
-						goto out;
-
-					if (exec) {
-						/*
-						 * If we're in ExecSync we don't output the k8s log
-						 * format. TODO(cyphar): This code really should be
-						 * rewritten so that we have a single conmon per
-						 * container and the conmon is logging the main
-						 * container process as a separate piece of logic to
-						 * the streaming to Exec[Sync] clients.
-						 */
-						if (write(logfd, buf, num_read) < 0) {
-							nwarn("write failed");
-							goto out;
-						}
-					} else {
-						if (write_k8s_log(logfd, "stdout", buf, num_read) < 0) {
-							nwarn("write_k8s_log failed");
-							goto out;
-						}
-					}
+				int masterfd = evlist[i].data.fd;
+				stdpipe_t pipe;
+				if (masterfd == masterfd_stdout)
+					pipe = STDOUT_PIPE;
+				else if (masterfd == masterfd_stderr)
+					pipe = STDERR_PIPE;
+				else {
+					nwarn("unknown pipe fd");
+					goto out;
+				}
+
+				num_read = read(masterfd, buf, BUF_SIZE);
+				if (num_read <= 0)
+					goto out;
+
+				if (write_k8s_log(logfd, pipe, buf, num_read) < 0) {
+					nwarn("write_k8s_log failed");
+					goto out;
 				}
 			} else if (evlist[i].events & (EPOLLHUP | EPOLLERR)) {
 				printf("closing fd %d\n", evlist[i].data.fd);
 				if (close(evlist[i].data.fd) < 0)
 					pexit("close");
-				goto out;
+				num_stdio_fds--;
```

Comment:
If you're referring to runC's code I'd say this is fine because `epoll` won't send a close on a file descriptor that's already closed. But I can also remove it from the `epollfd` as well.